### PR TITLE
Add decorators to events, templates chapters.

### LINKED ISSUE #886: Documentation on accessing elements in the template
I know that shadowRoot is DOM standard but it may increase acceptance if you add some notes about accessing elements in the template to the documentation. I was converting a Polymer app to lit-element and got stuck at the following line:

`this.$.myDialog.open()`

It took me a while to find out that there is shadowRoot in the DOM which would allow me the following:

`this.shadowRoot.getElementById("myDialog").open();`

I found a line like this in an anti-pattern example in the templates chapter. But in this case I am not setting a property but calling a method of a Polymer paper-dialog. I hope this makes sense?

### PR DESCRIPTION
**Pull request by arthurevans**
Decorator docs part 2/2.

Follow-on to #999.
Fixes #886.

Staged as: https://decorator-docs-2-dot-polymer-lit-element.uk.r.appspot.com

(Note that this branch doesn't include the new decorators chapter--that's in #999--so the links to that chapter are broken in the staging link above.)

### CODE REVIEW

**Comment by arthurevans** [comment]
Updated staging with the latest commits. PTAL.

**Comment by arthurevans** [comment]
@justinfagnani @sorvell could one of you take a look at this one?

File: docs/_guide/events.md [hunk 1]
```diff
@@ -8,96 +8,156 @@ slug: events
...
+render() { 
+  return html`
+    <button @click=${this._handleClick}>Click me!</button>
+  `;
+}
+```
+
+<div class="alert alert-info">
+
+**Using decorators.** Decorators are a proposed JavaScript feature, so you’ll need to use a compiler like Babel or the TypeScript compiler to use decorators. See Using decorators for details.
```
Comment (justinfagnani) [inline docs/_guide/events.md:147]: make "Using decorators" a link

Comment (justinfagnani) [inline docs/_guide/events.md:147]: maybe?

```suggestion
**Using decorators.** Decorators are a proposed JavaScript feature, so you’ll need to use a compiler like Babel or TypeScript to use decorators. See Using decorators for details.
```

File: docs/_guide/events.md [hunk 2]
```diff
@@ -8,96 +8,156 @@ slug: events
...
-    }
-    ```
+See [firstUpdated](/guide/lifecycle#firstupdated) in the Lifecycle documentation for more information.
+
 
-### Using `this` in event handlers
+## Using `this` in event listeners
 
-The default JavaScript context object (`this`) inside an event handler belonging to a LitElement-based component is the component itself. 
+Event listeners added using the declarative (`@event`) syntax in the template are automatically _bound_ to the component.
```
Comment (justinfagnani) [inline docs/_guide/events.md:97]: You should probably keep the "bound" verbiage here, but technically we don't bind the function, we just call it with the right `this` value. The difference is that we don't create a new bound function for each event binding * element instance, but just reuse the prototype type method. It's only a thing because people often criticize React for creating so many closures.

File: docs/_guide/events.md [hunk 3]
```diff
@@ -8,96 +8,156 @@ slug: events
...
+  _handleClick(e) {
     console.log(this.prop);
   }
 }
 ```
 
+When adding listeners imperatively with `addEventListener`, you'll need to bind the event listener yourself if you need a reference to the component instance. For example:
+
+```js
+this.boundResizeHandler = this.handleResize.bind(this);
```
Comment (justinfagnani) [inline docs/_guide/events.md:115]: I think the most ergonomic way to do this is like:

```ts
class MyElement extends LitElement {
  private _onResize = (e) => { ... };

  constructor() {
    super();
    window.addEventListener('resize', this._onResize);
  }
}
```

The only downside here is you can't call `super._onResize()` in the handler, but that's a rare thing to need to do.

File: docs/_guide/events.md [hunk 4]
```diff
@@ -8,96 +8,156 @@ slug: events
...
+
 See the [documentation for `this` on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this) for more information.
 
-### Use cases
+## Setting event listener options
+
+When you add an event listener imperatively, using `addEventListener`, you can specify various event listener options. For example, to use a capture phase event listener in plain JavaScript you'd do something like this:
+
+```js
+someElement.addEventListener('click', this._handleClick, {capture: true});
```
Comment (justinfagnani) [inline docs/_guide/events.md:126]: I would suggest using a `'scroll'` event and the `passive: true` option just 'cause more people should know about it (it eliminates jank from scroll listeners).

File: docs/_guide/templates.md [hunk 5]
```diff
@@ -375,21 +386,22 @@ You can compose LitElement templates from other LitElement templates. In the fol
 
 ```js
 class MyPage extends LitElement {
+  ...
   render() {
     return html`
-      ${this.headerTemplate}
-      ${this.articleTemplate}
+      ${this.headerTemplate(this.article.title)}
```
Comment (justinfagnani) [inline docs/_guide/templates.md:392]: Instead of passing an argument I would just access `this.article.title` in the headerTemplate. Or keep the argument and make header template standalone function to show that they don't need to be class methods.

File: docs/_guide/templates.md [hunk 6]
```diff
@@ -569,6 +581,162 @@ html`${until(content, html`<span>Loading...</span>`)}`
 
 For a list of directives supplied with lit-html, see [Built-in directives](https://lit-html.polymer-project.org/guide/template-reference#built-in-directives) in the Template syntax reference.
 
+## Accessing nodes in the shadow DOM
+
+Templated DOM is usually rendered into shadow DOM, so the nodes are not direct children of the component. Use `this.shadowRoot.querySelector()` or `this.shadowRoot.querySelectorAll()` to find nodes in the
```
Comment (justinfagnani) [inline docs/_guide/templates.md:586]: maybe?

```suggestion
The `render()` method result is usually rendered into shadow DOM, so the nodes are not direct children of the component. Use `this.shadowRoot.querySelector()` or `this.shadowRoot.querySelectorAll()` to find nodes in the 
```

only because "templated DOM" seems to be a phrase of our own creation. I might change "templated DOM" below to "rendered DOM" too.

File: docs/_guide/templates.md [hunk 7]
```diff
@@ -569,6 +581,162 @@ html`${until(content, html`<span>Loading...</span>`)}`
...
+
+**Using decorators.** Decorators are a proposed JavaScript feature, so you’ll need to use a compiler like Babel or the TypeScript compiler to use decorators. See [Using decorators](decorators) for details.
+
+</div>
+
+The `@query` decorator modifies a class property, turning it into a getter that returns a node from the render root.
+
+```js
+class MyElement {
+  @query('#first')
```
Comment (justinfagnani) [inline docs/_guide/templates.md:618]: I would suggest making this a TypeScript example, making the field private, and naming it with a `_` prefix, because the queried elements are internal implementation details and `@query` is really meant to internal usage - we don't want authors accidentally making private things public.

File: docs/_guide/templates.md [hunk 8]
```diff
@@ -569,6 +581,162 @@ html`${until(content, html`<span>Loading...</span>`)}`
...
 
+## Accessing nodes in the shadow DOM
+
+Templated DOM is usually rendered into shadow DOM, so the nodes are not direct children of the component. Use `this.shadowRoot.querySelector()` or `this.shadowRoot.querySelectorAll()` to find nodes in the 
+shadow DOM.
+
+You can query the templated DOM after its initial render (for example, in `firstUpdated`), or use a getter pattern, like this:
+
+```js
+get closeButton() {
```
Comment (justinfagnani) [inline docs/_guide/templates.md:592]: Same as the comment on the `@query` example below, I would emphasize that this should probably be "private" unless you really want to expose the element publicly. So name it with a `_` prefix.

File: docs/_guide/templates.md [hunk 9]
```diff
@@ -569,6 +581,162 @@ html`${until(content, html`<span>Loading...</span>`)}`
...
+```
+@queryAll('button')
+buttons!: NodeListOf<HTMLButtonElement>
+```
+
+The exclamation point (`!`) after `buttons` is TypeScript's [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator). It tells the compiler to treat `buttons` as always being defined, never `null` or `undefined`.
+
+Finally, `@queryAsync` works like `@query`, except that instead of returning a node directly, it returns a `Promise` that resolves to that node. External code can use this instead of waiting for the `updateComplete` promise. 
+
+This is useful, for example, if the node returned by `@queryAsync` can change as a result of another property change.
```
Comment (justinfagnani) [inline docs/_guide/templates.md:672]: Maybe @sorvell can add some insight on documenting `@queryAsync` externally. It's most for internal tool usage, and I'm a little worried that it seems like we're encouraging exposing shadow DOM contents publicly here. I wonder how we can describe this well, or if we should just ignore `@queryAsync` in the main docs.

File: docs/_guide/templates.md [hunk 10]
```diff
@@ -569,6 +581,162 @@ html`${until(content, html`<span>Loading...</span>`)}`
...
+*   [HTMLSlotElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement) on MDN.
+
+
+### @queryAssignedNodes decorator
+
+The `@queryAssignedNodes` decorator converts a class property into a getter that returns all of the assigned nodes for a given slot in the component's shadow tree. 
+
+<div class="alert alert-info">
+
+**Using decorators.** Decorators are a proposed JavaScript feature, so you’ll need to use a compiler like Babel or the TypeScript compiler to use decorators. See [Using decorators](decorators) for details.
```
Comment (justinfagnani) [inline docs/_guide/templates.md:715]: Is there a way to define this snippet in a common location and include it?

File: docs/_guide/templates.md [hunk 11]
```diff
@@ -569,6 +581,162 @@ html`${until(content, html`<span>Loading...</span>`)}`
...
+
+</div>
+
+```js
+// First argument is the slot name
+// Second argument is `true` to flatten the assigned nodes.
+@queryAssignedNodes('header', true)
+headerNodes;
```
Comment (justinfagnani) [inline docs/_guide/templates.md:723]: Maybe make these private too, though it's a little less of a concern because child nodes are already public external to the element. I just want to reinforce that class members defined only to help with the component implementation should be private by default.

File: docs/_includes/projects/docs/templates/compose/my-page.js [hunk 12]
```diff
@@ -1,21 +1,36 @@
...
+      text: 'Some witty text.',
+    };
+  }
+
   render() {
     return html`
-      ${this.headerTemplate}
-      ${this.articleTemplate}
-      ${this.footerTemplate}
+      ${this.headerTemplate(this.article.title)}
```
Comment (justinfagnani) [inline docs/_includes/projects/docs/templates/compose/my-page.js:22]: Same comment with the argument as above.

File: docs/_guide/events.md (hunk 4 above)
Comment (arthurevans) [reply docs/_guide/events.md:126]: My understanding was that `scroll` itself can't be canceled, so it doesn't need `passive`? I think you'd use it on `touchstart`, `touchend`, etc... https://developers.google.com/web/updates/2016/06/passive-event-listeners

File: docs/_guide/templates.md (hunk 5 above)
Comment (arthurevans) [reply docs/_guide/templates.md:392]: Urgh. I was trying to show functions here because someone was confused by the getters. 

Are standalone methods too confusing? I like showing that they don't need to be class methods, so I'll go with that.

File: docs/_guide/events.md (hunk 1 above)
Comment (arthurevans) [reply docs/_guide/events.md:147]: Done.

Comment (arthurevans) [reply docs/_guide/events.md:147]: Updated this in all the places it appears.

File: docs/_guide/templates.md (hunk 10 above)
Comment (arthurevans) [reply docs/_guide/templates.md:715]: Possibly? I think we'll only end up with a total of 4 or 5 instances of it in the doc set, though.

File: docs/_guide/templates.md (hunk 11 above)
Comment (arthurevans) [reply docs/_guide/templates.md:723]: Sure, SG.

File: docs/_guide/events.md (hunk 3 above)
Comment (arthurevans) [reply docs/_guide/events.md:115]: There is also a certain weird tax. I've seen this before and I still had to read a half-dozen pages of doc to remind myself why it works.

Also: assumes class fields support. Should I include both examples, perhaps?

File: docs/_guide/events.md (hunk 2 above)
Comment (arthurevans) [reply docs/_guide/events.md:97]: OK, so just for the record--I'm _not_ changing any doc here, right?

File: docs/_guide/templates.md (hunk 6 above)
Comment (arthurevans) [reply docs/_guide/templates.md:586]: It *is* a phrase of our own creation. We already use it a couple of times at the beginning of the chapter.

In the lit-html docs, we actually just say "templates"--like, "Styling templates"--when we really mean, "styling the DOM you create by rendering templates." If you think "rendered DOM" is clearer and less invented-y, we could rephrase this to, say, "Rendered DOM is usually placed in an open shadow root," or "The component's template is usually rendered into shadow DOM" and use "rendered DOM" below as you suggested.

Added this to today's docs meeting agenda if we need to discuss it further.

File: docs/_guide/templates.md (hunk 8 above)
Comment (arthurevans) [reply docs/_guide/templates.md:592]: So named.

File: docs/_guide/templates.md (hunk 7 above)
Comment (arthurevans) [reply docs/_guide/templates.md:618]: At the moment, we don't have anything to show it's a Typescript example, so I've been trying to limit Typescript-specific syntax to examples that are explicitly labeled Typescript (one more reason I'm dying to get a working code-switcher up and running).

How about I just use the `_` and add a note about how these properties should typically be protected or private (`_` convention or Typescript's `private` or `protected` keywords).

File: docs/_guide/templates.md (hunk 9 above)
Comment (arthurevans) [reply docs/_guide/templates.md:672]: By "internal tool usage" do you mean for testing? I'm not actually too clear on the use cases for this decorator. 

Help me @sorvell you're my only hope!

File: docs/_includes/projects/docs/templates/compose/my-page.js (hunk 12 above)
Comment (arthurevans) [reply docs/_includes/projects/docs/templates/compose/my-page.js:22]: Fixed, thanks!

File: docs/_guide/events.md (hunk 4 above)
Comment (arthurevans) [reply docs/_guide/events.md:126]: Used "touchstart" with passive.

File: docs/_guide/events.md (hunk 3 above)
Comment (arthurevans) [reply docs/_guide/events.md:115]: Included both.

File: docs/_guide/templates.md (hunk 9 above)
Comment (sorvell) [reply docs/_guide/templates.md:672]: Sorry, that comment was unclear.

I was imagining the use case where a dev decides to expose a public property via `@query('.foo') fooElement`. In order for that to work correctly at all times, the user would have to `await element.updateComplete`:

```
await element.updateComplete;
element.fooElement.textContent = 'hi';
```
Instead the dev may choose to expose the property with `@queryAsync('.foo') fooElement;` so the user code can be more compact:

```
(await element.fooElement).textContent = 'hi';
```

Comment (arthurevans) [reply docs/_guide/templates.md:678]: OK, understood. I think the concern @justinfagnani had was that these properties that expose shadow DOM children should probably be internal so we don't leak implementation details.

That said, a property getter provides more controlled access than having the user code call `element.shadowRoot.querySelector()` themselves. So...

Comment (arthurevans) [reply docs/_guide/templates.md:678]: For now, just leaving this in with the word "external" removed.
